src/llvm/mod.rs

```diff
@@ -179,69 +181,130 @@ pub unsafe fn optimize(
+        OptLevel::No | OptLevel::Less => "default<O1>",
+        OptLevel::Default => "default<O2>",
+        OptLevel::Aggressive => "default<O3>",
+        OptLevel::Size => "default<Os>",
+        OptLevel::SizeMin => "default<Oz>",
+    })
+    .unwrap();
+
+    debug!("running optimization pass");
+    LLVMRunPasses(module, opt.as_ptr(), tm, options);
+
+    LLVMDisposePassBuilderOptions(options);
 
     // Some debug info generated by rustc seems to trigger a segfault in the
     // BTF code in llvm, so strip it until that is fixed
     LLVMStripModuleDebugInfo(module);
+
+    // Collect up all the memory intrinsics. We're going to replace these with
+    // calls to the equivalent functions, but we can't remove the instructions
+    // until we're done iterating over them.
+    let mut mem_intrinsic_instructions = Vec::new();
+    for function in module.functions_iter() {
+        for basic_block in function.basic_blocks_iter() {
+            for instruction in basic_block.instructions_iter() {
+                let instruction = LLVMIsAMemIntrinsic(instruction);
+                if instruction.is_null() {
+                    continue;
+                }
+                mem_intrinsic_instructions.push(instruction);
+            }
+        }
+    }
+
+    // LLVMGet* functions do not pass ownership to the caller.
+    let context = LLVMGetModuleContext(module);
+
+    let builder = LLVMCreateBuilderInContext(context);
+    for instruction in mem_intrinsic_instructions.iter().copied() {
+        let instruction_num_operands = LLVMGetNumOperands(instruction);
+        let instruction_num_operands: u32 = instruction_num_operands.try_into().unwrap();
+        let mut instruction_operands: Vec<_> = (0..instruction_num_operands)
+            .map(|i| LLVMGetOperand(instruction, i))
+            .collect();
+
+        let instruction_name = if !LLVMIsAMemCpyInst(instruction).is_null() {
+            "memcpy"
+        } else if !LLVMIsAMemMoveInst(instruction).is_null() {
+            "memmove"
+        } else if !LLVMIsAMemSetInst(instruction).is_null() {
+            "memset"
+        } else {
+            panic!("unknown mem intrinsic");
+        };
+        let function = {
+            let instruction_name = CString::new(instruction_name).unwrap();
+            LLVMGetNamedFunction(module, instruction_name.as_ptr())
+        };
+        // The user neglected to defined a required intrinsic. Perhaps we
+        // should return an error instead of panicking.
+        assert!(
+            !function.is_null(),
+            "missing mem intrinsic function {instruction_name}"
+        );
+        let function_type = LLVMGlobalGetValueType(function);
+        let function_type = {
+            let return_type = LLVMGetReturnType(function_type);
+            let param_count = LLVMCountParamTypes(function_type);
+            // Use the parameter types we're about to pass rather than the parameter types of the function.
```

Comment:
Does this definitely work? Does LLVM do something smart to make sure the right data gets into the call?

Reply:
I don't know how to answer the second half of this question.

Take a look at the commit view on this PR - you'll see that adding this makes CI go from failing to passing.

Reply:
Well, we had a value that was one byte being passed to the intrinsic, and now for the builtin we're passing a value that is 4 bytes, but we're saying its type is one byte (right?). Are we just seeing it work because we're running on little-endian machines and we're passing the arguments through registers such that a 4-byte value is the same as a 1-byte value?

Reply:
Added an explicit cast.

Reply:
The code is cleaner this way, even if it was all correct without the cast. Thank you.